### PR TITLE
[Core] Add PointsNumberInDirection to Geometry

### PR DESCRIPTION
**Pull request by tteschemacher**
Hi all,

this function provides access to the number of points in each respective direction. For some geometries this might be known or computed in other ways. However, some come with dynamic sizes.

### CODE REVIEW

File: kratos/geometries/nurbs_surface_geometry.h
Comment (dagmawibekel) [inline kratos/geometries/nurbs_surface_geometry.h]: Doesn't it make sense to return an error or a warning, if the directionIndex doesn't match 1 or 2?

Comment (tteschemacher) [reply kratos/geometries/nurbs_surface_geometry.h]: Thanks for your comment! I am not sure. If you generalize a procedure with loops in all directions it would provide you the appropriate location in the list of nodes. Thus, I think if a user wants to avoid misuse it should ask about the dimensions of the geometries.

Comment (dagmawibekel) [reply kratos/geometries/nurbs_surface_geometry.h]: I think i didn't make myself clear enough. What i'm trying to say is that it might make sense to return a warning or an error instead of returning 1,  if the user input is wrong. For example, you are doing this already for the number of knotspans a couple lines below:

            KRATOS_ERROR << "NurbsSurfaceGeometry::NumberOfKnotSpans: Direction index: "
                << DirectionIndex << " not available. Options are: 0 and 1." << std::endl;

Furthermore i find the name of the function a little confusing. Maybe NumberOfPointsInDirection is a little better, but it's as above just a matter of stlye i guess. 

Cheers

Comment (tteschemacher) [reply kratos/geometries/nurbs_surface_geometry.h]: thanks @dagmawibekel I try to avoid that warning because what you could do is a generic method that iterates in all directions it iterates only 1 within the thickness direction. This is a legal operation and without lack of efficiency.

I am also not happy with the name, however, the corresponding function is called PointsNumber - that's why I tried to keep that.

**Review by dagmawibekel — DISMISSED** [review]
Sounds good to me

**Review by dagmawibekel — DISMISSED** [review]

**Comment by tteschemacher** [comment]
Thanks @dagmawibekel . I will wait for the approval of the @KratosMultiphysics/technical-committee until I merge.

File: kratos/geometries/geometry.h
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: It would be better to send an error here in the base class and only implement it in the ones that make sense.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee agrees with the new interface, but there are modifications to be done:
1. The base class should return an error and the virtual function should be implemented in the geometries giving this feature
2. The documentation and the argument name should indicate that the direction is in the local coordinates

File: kratos/geometries/quadrilateral_2d_4.h
Comment (rubenzorrilla) [inline kratos/geometries/quadrilateral_2d_4.h]: Shouldn't this be `LocalDirectionIndex > 1`?

File: kratos/geometries/nurbs_surface_geometry.h
Comment (rubenzorrilla) [inline kratos/geometries/nurbs_surface_geometry.h]: Same comment? The behaviour is undefined if the user provides `LocalDirectionIndex = 2`.

File: kratos/geometries/quadrilateral_2d_8.h
Comment (rubenzorrilla) [inline kratos/geometries/quadrilateral_2d_8.h]: Again. What happens if the user provides 2?

File: kratos/geometries/quadrilateral_2d_9.h
Comment (rubenzorrilla) [inline kratos/geometries/quadrilateral_2d_9.h]: Same.

**Review by rubenzorrilla — COMMENTED** [review]
Isn't the behaviour undefined for the  `LocalDirectionIndex = 2` argument?

**Comment by tteschemacher** [comment]
ah yes, sure. Thanks @rubenzorrilla for that comment.

**Comment by tteschemacher** [comment]
> 
> 
> Isn't the behaviour undefined for the `LocalDirectionIndex = 2` argument?

I changed it accordingly. Thanks for your comment.

Changing directly in github was in the end not the wisest idea...

**Comment by tteschemacher** [comment]
Can we merge this?

**Comment by tteschemacher** [comment]
@pooyan-dadvand , @rubenzorrilla can we merge this?

**Review by philbucher — APPROVED** [review]
Amazing!

**Comment by tteschemacher** [comment]
Beautiful ❤️

**Pull request merged**
